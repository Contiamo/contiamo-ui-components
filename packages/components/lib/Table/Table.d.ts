/// <reference types="react" />
import * as React from "react";
import { Css } from "../types";
export interface Props {
    css?: Css;
    className?: string;
    children?: React.ReactNode;
    /** Table columns headings */
    columns: string[];
    /** Table rows as an array of cells */
    rows: ((string | React.ReactNode)[])[];
    __experimentalColumnCss?: Css[];
    __experimentalRowActions?: React.ReactNode[];
    /** Called on row click */
    onRowClick?: (row: (string | React.ReactNode)[], index: number) => void;
}
declare const Table: (props: Props) => JSX.Element;
export default Table;
