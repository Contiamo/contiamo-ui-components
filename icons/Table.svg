<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 12.46">
  <defs>
    <style>
      .cls-1 {
        fill: #fff;
      }
    </style>
  </defs>
  <title>Table</title>
  <g id="Layer_2" data-name="Layer 2">
    <g id="Layer_1-2" data-name="Layer 1">
      <path class="cls-1" d="M15,0H1A1,1,0,0,0,0,1V11.46a1,1,0,0,0,1,1H15a1,1,0,0,0,1-1V1A1,1,0,0,0,15,0ZM7.5,3.16V6.81H1V3.16ZM1,7.81H7.5v3.65H1ZM15,12v-.5H8.5V7.81H15v3.65h0Zm0-5.15H8.5V3.16H15Z"/>
    </g>
  </g>
</svg>
