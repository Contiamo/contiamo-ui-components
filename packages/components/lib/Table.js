"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var glamorous_1 = require("glamorous");
var Container = glamorous_1.default.table(function (_a) {
    var theme = _a.theme;
    return ({
        label: "table",
        width: "100%",
        minWidth: 320,
        borderCollapse: "collapse",
        textAlign: "left",
        backgroundColor: "white",
        "& th": __assign({}, theme.typography.body, { opacity: 0.4 }),
        "& tr": {
            borderTop: "1px solid",
            borderBottom: "1px solid"
        },
        "& tr:first-child": {
            borderTop: 0
        },
        "& tbody tr:last-child": {
            borderBottom: 0
        },
        "& td": __assign({ padding: theme.spacing * 4 / 3 + "px " + theme.spacing + "px" }, theme.typography.body),
        "& td, & th, & tr": {
            borderColor: theme.colors.separator,
            padding: theme.spacing / 2 + "px " + theme.spacing + "px"
        },
        "& tfoot": {
            height: 2 * theme.spacing,
            position: "relative"
        }
    });
});
var EmptyViewTFoot = glamorous_1.default.tfoot(function (_a) {
    var theme = _a.theme;
    return ({
        width: "100%",
        height: 2 * theme.spacing,
        position: "relative"
    });
});
var EmptyView = glamorous_1.default.span(function (_a) {
    var theme = _a.theme;
    return (__assign({ display: "inline-block", width: "100%", position: "absolute", top: theme.spacing, left: "50%", transform: "translate3d(-50%, -50%, 0)" }, theme.typography.body));
});
var Table = function (props) { return (React.createElement(Container, { css: props.css, className: props.className },
    React.createElement("thead", null,
        React.createElement("tr", null, props.columns.map(function (column, index) { return React.createElement("th", { key: index }, column); }))),
    React.createElement("tbody", null, props.rows.map(function (row, index) { return React.createElement("tr", { key: index }, row.map(function (cell, index) { return React.createElement("td", { key: index }, cell); })); })),
    props.rows.length === 0 ? (React.createElement(EmptyViewTFoot, null,
        React.createElement(EmptyView, null, "This table is empty"))) : null)); };
exports.default = Table;
//# sourceMappingURL=Table.js.map