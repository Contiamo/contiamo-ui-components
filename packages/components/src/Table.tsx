import * as React from "react"
import glamorous from "glamorous"
import { Theme } from "@operational/theme"

export interface Props {
  css?: {}
  className?: string
  children?: React.ReactNode
  columns: string[]
  rows: ((string | React.ReactNode)[])[]
}

const Container = glamorous.table(({ theme }: { theme: Theme }): {} => ({
  label: "table",
  border: 0,
  borderCollapse: "collapse",
  textAlign: "left",
  backgroundColor: "white",
  "& th": {
    border: "1px solid",
    ...theme.typography.body,
    fontWeight: 600
  },
  "& td": {
    border: "1px solid",
    ...theme.typography.body
  },
  "& td p": {
    marginBottom: 0
  },
  "& td, & th": {
    borderColor: theme.colors.gray20,
    padding: `${theme.spacing / 2}px ${theme.spacing}px`
  }
}))

const Table = (props: Props) => (
  <Container css={props.css} className={props.className}>
    <thead>
      <tr>{props.columns.map((column, index) => <th key={index}>{column}</th>)}</tr>
    </thead>
    <tbody>
      {props.rows.map((row, index) => <tr key={index}>{row.map((cell, index) => <td key={index}>{cell}</td>)}</tr>)}
    </tbody>
  </Container>
)

export default Table
