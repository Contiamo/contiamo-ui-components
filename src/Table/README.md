Tables simply render a semantic HTML table structure based on raw data.

### Simple usage

You can render a simple use-case of the table by specifying a list of records and supplying the columns as a list of
strings which would serve both as column heading and access keys for fields.

```jsx
import * as React from "react"
import { Table } from "@operational/components"
;<Table
  data={[{ name: "Max", profession: "Carpenter" }, { name: "Moritz", profession: "Baker" }]}
  columns={["name", "profession"]}
/>
```

### With Fixed Layout

Tables perform better with a forced fixed layout, since the browser doesn't have to recalculate positions depending on
the contents of the table. Here's what the same table looks like with a fixed layout.

```jsx
import * as React from "react"
import { Table } from "@operational/components"
;<Table
  fixedLayout
  data={[{ name: "Max", profession: "Carpenter" }, { name: "Moritz", profession: "Baker" }]}
  columns={["name", "profession"]}
/>
```

### With an "active" row

Tables perform better with a forced fixed layout, since the browser doesn't have to recalculate positions depending on
the contents of the table. Here's what the same table looks like with a fixed layout.

```jsx
import * as React from "react"
import { Table } from "@operational/components"
;<Table
  fixedLayout
  activeRowIndex={1}
  data={[
    { name: "Max", profession: "Carpenter" },
    { name: "Michael", profession: "Anti-focusring activist" },
    { name: "Moritz", profession: "Baker" },
  ]}
  columns={["name", "profession"]}
/>
```

### With Reorder capability

```jsx
import * as React from "react"
import { Table } from "@operational/components"

function reorder<T>(list: T[], startIndex: number, endIndex: number) {
  const result = Array.from(list)
  const [removed] = result.splice(startIndex, 1)
  result.splice(endIndex, 0, removed)

  return result
}

const MyComponent = () => {
  const [data, setData] = React.useState([
    { name: "Max", profession: "Carpenter" },
    { name: "Moritz", profession: "Baker" },
    { name: "Tejas", profession: "Useless" },
    { name: "Fabien", profession: "Genius" },
    { name: "Mischa", profession: "Wise One" },
    { name: "Imogen", profession: "Math Guru" },
    { name: "Slava", profession: "Stereo Booster" },
  ])

  return (
    <div style={{ height: 500 }}>
      <Table
        fixedLayout
        data={data}
        onReorder={result => {
          alert("Dropped!")
          setData(reorder(data, result.source.index, result.destination.index))
        }}
        columns={["name", "profession"]}
      />
    </div>
  )
}

;<MyComponent />
```

### Simple Usage without Header

```jsx
import * as React from "react"
import { Table } from "@operational/components"
;<Table
  headless
  data={[{ name: "Max", profession: "Carpenter" }, { name: "Moritz", profession: "Baker" }]}
  columns={["name", "profession"]}
/>
```

While this approach is convenient, it is not recommended because it makes it too easy to re-use record keys as table
headings, and adds a strong coupling between data and view concerns.

We suggest taking the time to think about the best way to describe the data fields and then specifying it explicitly as
column headers. The following, slightly more verbose version of the API demonstrates how this can be achieved:

### With explicit data formatting

In this case, simple functions taking an individual record as an argument specify how cells should be rendered in a
given column:

```jsx
import * as React from "react"
import { Table, OlapIcon, Chip, AvatarGroup } from "@operational/components"

const data = [
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Toy Bundle",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
]
;<Table
  data={data}
  columns={[
    { heading: "", cell: dataEntry => <OlapIcon color="primary" /> },
    { heading: "Name", cell: dataEntry => dataEntry.name },
    { heading: "Last updated", cell: dataEntry => dataEntry.lastUpdated },
    { heading: "Tags", cell: dataEntry => dataEntry.tags.map((tag, tagIndex) => <Chip key={tagIndex}>{tag}</Chip>) },
    { heading: "Collaborators", cell: dataEntry => <AvatarGroup avatars={dataEntry.collaborators} /> },
  ]}
  onRowClick={(dataEntry, i) => console.log({ dataEntry, i })}
/>
```

### With custom widths per column

We can customize the look of the table by specifying a `width` property on certain columns.

```jsx
import * as React from "react"
import { Table, Chip, AvatarGroup, OlapIcon } from "@operational/components"

const data = [
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Toy Bundle",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
]
;<Table
  data={data}
  columns={[
    { heading: "", cell: dataEntry => <OlapIcon color="primary" /> },
    { heading: "Name", width: 600, cell: dataEntry => dataEntry.name },
    { heading: "Last updated", cell: dataEntry => dataEntry.lastUpdated },
    {
      heading: "Tags",
      width: 100,
      cell: dataEntry => dataEntry.tags.map((tag, tagIndex) => <Chip key={tagIndex}>{tag}</Chip>),
    },
    { heading: "Collaborators", cell: dataEntry => <AvatarGroup avatars={dataEntry.collaborators} /> },
  ]}
  onRowClick={(dataEntry, i) => console.log({ dataEntry, i })}
/>
```

### With icons

```jsx
import * as React from "react"
import { Table, ProjectIcon, OrganizationIcon, Chip } from "@operational/components"

const data = [
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
  {
    name: "Toy Bundle",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
]
;<Table
  data={data}
  icon={dataEntry => OrganizationIcon}
  columns={[
    { heading: "", cell: dataEntry => <ProjectIcon color="primary" /> },
    { heading: "Name", cell: dataEntry => dataEntry.name },
    { heading: "Last updated", cell: dataEntry => dataEntry.lastUpdated },
    { heading: "Tags", cell: dataEntry => dataEntry.tags.map((tag, tagIndex) => <Chip key={tagIndex}>{tag}</Chip>) },
  ]}
  onRowClick={(dataEntry, i) => console.log({ dataEntry, i })}
/>
```

### With default row action

```jsx
import * as React from "react"
import { Table, OrganizationIcon, ProjectIcon, Chip } from "@operational/components"

const data = [
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
  {
    name: "Toy Bundle",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
  },
]
;<Table
  data={data}
  icon={dataEntry => OrganizationIcon}
  iconColor={dataEntry => (dataEntry.name.indexOf("Mega") > -1 ? "primary" : "gray")}
  columns={[
    { heading: "", cell: dataEntry => <ProjectIcon color="primary" /> },
    { heading: "Name", cell: dataEntry => dataEntry.name },
    { heading: "Last updated", cell: dataEntry => dataEntry.lastUpdated },
    { heading: "Tags", cell: dataEntry => dataEntry.tags.map((tag, tagIndex) => <Chip key={tagIndex}>{tag}</Chip>) },
  ]}
  onRowClick={(dataEntry, i) => console.log({ dataEntry, i })}
  rowActionName="View Details"
/>
```

### With row actions

Row actions are specified as a function of an individual record, returning action items conforming to the
[ContextMenu](/#ContextMenu) API. The function in the `rowActions` prop can return either the action items as an array
or the `ActionMenu` node itself.

```jsx
import * as React from "react"
import { Table, Chip, AvatarGroup, FunctionIcon } from "@operational/components"

const data = [
  {
    name: "Mega Deal Dev",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Small deal prod",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
  {
    name: "Toy Bundle",
    lastUpdated: "2018-06-06",
    tags: ["agent-view", "production"],
    collaborators: [
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
      { photo: "https://graph.facebook.com/775278331/picture", name: "Tejas Kumar" },
    ],
  },
]
;<Table
  data={data}
  columns={[
    { heading: "", cell: dataEntry => <FunctionIcon color="primary" /> },
    { heading: "Name", cell: dataEntry => dataEntry.name },
    { heading: "Last updated", cell: dataEntry => dataEntry.lastUpdated },
    { heading: "Tags", cell: dataEntry => dataEntry.tags.map((tag, tagIndex) => <Chip key={tagIndex}>{tag}</Chip>) },
    { heading: "Collaborators", cell: dataEntry => <AvatarGroup avatars={dataEntry.collaborators} /> },
  ]}
  onRowClick={(dataEntry, dataEntryIndex) => alert("You chose " + dataEntry.name)}
  rowActions={dataEntry => [
    {
      label: "Details",
      onClick: () => {
        alert("Details on " + dataEntry.name)
      },
    },
    {
      label: "Delete",
      onClick: () => {
        alert("Deleting " + dataEntry.name)
      },
    },
  ]}
/>
```

### Without data

Tables render a default empty view if no records are specified.

```jsx
import * as React from "react"
import { Table } from "@operational/components"
;<Table columns={["", "Name", "Last updated", "Tags", "Collaborators"]} data={[]} />
```

### With sorting

Tables can be sorted by custom logic, it's totally up to you to implement the sort logic that you need.

```jsx
import * as React from "react"
import { Table, Chip } from "@operational/components"

const MyComponent = () => {
  const [data, setData] = React.useState([
    { name: "Imogen", tags: ["d3js"] },
    { name: "Tejas", tags: ["lambda"] },
    { name: "Fabien", tags: ["regex"] },
    { name: "Peter", tags: ["webGL"] },
  ])
  const [nameSortOrder, setNameSortOrder] = React.useState(undefined)
  const [tagSortOrder, setTagSortOrder] = React.useState(undefined)

  const sortData = (order, key) => {
    if (key === "name") {
      setData(
        data.sort((a, b) => {
          if (order === "asc") return a.name < b.name ? -1 : 1
          return a.name > b.name ? -1 : 1
        }),
      )
      setNameSortOrder(order)
      setTagSortOrder(undefined)
    } else {
      setData(
        data.sort((a, b) => {
          if (order === "asc") return a.tags[0] < b.tags[0] ? -1 : 1
          return a.tags[0] > b.tags[0] ? -1 : 1
        }),
      )
      setNameSortOrder(undefined)
      setTagSortOrder(order)
    }
  }
  return (
    <Table
      data={data}
      columns={[
        {
          heading: "Name",
          cell: i => i.name,
          sortOrder: nameSortOrder,
          onSortClick: order => sortData(order, "name"),
        },
        {
          heading: "Tags",
          cell: i => i.tags.map(t => <Chip key={t}>{t}</Chip>),
          sortOrder: tagSortOrder,
          onSortClick: order => sortData(order, "tag"),
        },
      ]}
    />
  )
}

;<MyComponent />
```
