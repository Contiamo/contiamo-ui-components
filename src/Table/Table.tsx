import * as React from "react"
import {
  Droppable,
  Draggable,
  DragDropContext,
  ResponderProvided,
  DropResult,
  DraggableProps,
} from "react-beautiful-dnd"

import ActionMenu, { ActionMenuProps } from "../ActionMenu/ActionMenu"
import { DefaultProps } from "../types"
import Small from "../Typography/Small"
import styled from "../utils/styled"
import { IconComponentType, ChevronDownIcon, ChevronUpDownIcon, ChevronUpIcon } from "../Icon"
import { useUniqueId } from "../useUniqueId"
import { lighten } from "../utils"

export interface TableProps<T> extends DefaultProps {
  data: T[]
  /** Table columns headings */
  columns: Array<Column<T> | Extract<keyof T, string>>
  /** Called on row click */
  onRowClick?: (dataEntry: T, index: number) => void
  /** Label to show on row hover */
  rowActionName?: string
  /**
   * Add actions on the end of each row
   */
  rowActions?: (dataEntry: T) => ActionMenuProps["items"] | React.ReactNode
  /** Icon name for row */
  icon?: (dataEntry: T) => IconComponentType
  /** Icon color for row */
  iconColor?: (dataEntry: T) => string
  /** Remove the header? */
  headless?: boolean
  /** Fixed Layout for performance */
  fixedLayout?: boolean
  /** On reorder rows, */
  onReorder?: (result: DropResult, provided: ResponderProvided) => void
  /* The index of an active row */
  activeRowIndices?: number[]
  /* Compact style rows */
  condensed?: boolean
}

export interface Column<T> {
  heading: React.ReactNode
  cell: (dataEntry: T, index: number) => React.ReactNode
  sortOrder?: "asc" | "desc"
  onSortClick?: (order: "asc" | "desc") => void
  width?: number
}

const CondensedActionMenu = styled(ActionMenu)<{ condensed?: boolean }>`
  ${({ condensed }) =>
    condensed
      ? `& div {
          margin-bottom: 0;
          height: 33px;
        }`
      : ""}
`

const Container = styled.table<{ fixedLayout: TableProps<any>["fixedLayout"] }>(({ theme, fixedLayout }) => ({
  width: "100%",
  backgroundColor: theme.color.white,
  textAlign: "left",
  borderCollapse: "collapse",
  fontSize: theme.font.size.small,
  fontFamily: theme.font.family.main,
  tableLayout: fixedLayout ? "fixed" : "initial",
}))

const Tr = styled.tr<{
  active: boolean
  isDragging?: boolean
  draggable?: boolean
  clickable?: boolean
  condensed?: boolean
}>(({ isDragging, theme, active, draggable, clickable, condensed }) => ({
  height: condensed ? 36 : 50,
  display: isDragging ? "table" : "table-row",
  tableLayout: "fixed",
  backgroundColor: active ? lighten(theme.color.primary, 54) : theme.color.white,
  ...(draggable || clickable
    ? {
        ":hover": {
          backgroundColor: active ? lighten(theme.color.primary, 52) : lighten(theme.color.primary, 54),
          cursor: clickable ? "pointer" : draggable ? "move" : "default",
        },
      }
    : {}),
  ":focus": {
    outline: "none",
    backgroundColor: lighten(theme.color.primary, 52),
  },
  ".no-focus &:focus": {
    backgroundColor: active ? lighten(theme.color.primary, 54) : theme.color.white,
  },
}))

const Thead = styled.thead`
  tr {
    height: initial;
  }
`

const Th = styled.th<{ sortable?: boolean }>(({ theme, sortable }) => ({
  position: "relative",
  borderBottom: `1px solid ${theme.color.separators.default}`,
  color: theme.color.text.dark,
  paddingBottom: theme.space.base,
  "&:first-of-type": {
    paddingLeft: theme.space.small,
  },
  paddingRight: theme.space.small,
  ...(sortable
    ? {
        ":hover": {
          cursor: "pointer",
          color: theme.color.text.light,
          svg: {
            cursor: "pointer",
            fill: theme.color.text.light,
          },
        },
      }
    : {}),
}))

const ThContent = styled.span<{ sorted?: boolean }>`
  display: inline-flex;
  align-items: center;
  height: ${props => props.theme.space.medium}px;
  ${props => props.sorted && `color: ${props.theme.color.text.light};`};
`

const Td = styled.td<{ cellWidth?: Column<any>["width"]; coloredBorders: boolean }>(
  ({ theme, cellWidth, coloredBorders }) => ({
    verticalAlign: "middle",
    borderBottom: "1px solid",
    color: theme.color.text.default,
    borderTop: "1px solid",
    borderColor: coloredBorders ? theme.color.primary : theme.color.border.medium,
    hyphens: "auto",
    "&:first-of-type": {
      paddingLeft: theme.space.small,
    },
    paddingRight: theme.space.small,
    ...(cellWidth
      ? {
          width: cellWidth,
          wordBreak: "break-all",
          wordWrap: "break-word",
        }
      : {}),
  }),
)

const Actions = styled(Td)(({ theme }) => ({
  textAlign: "right",
  paddingRight: theme.space.small,

  /**
   * We use opacity here instead of display: none; or
   * visibility: hidden; because both mess with
   * the box model of the Td while opacity does not.
   */
  opacity: 0,
  "tr:hover &, :hover, tr:focus &, :focus-within, &:focus": {
    opacity: 1,
  },

  "& > div": {
    display: "inline-flex",
  },
}))

const CellIcon = styled(Td)<{ condensed?: boolean }>`
  width: ${({ condensed }) => (condensed ? 28 : 42)}px;
  padding: ${({ theme }) => theme.space.base}px;
  color: ${({ theme }) => theme.color.text.lightest};
`

const ActionLabel = styled(Small)`
  color: ${props => props.theme.color.primary};
  margin: 0;
  display: block;
`

const EmptyView = styled(Td)<{ condensed?: boolean }>`
  color: ${({ theme }) => theme.color.text.default};
  height: ${({ condensed }) => (condensed ? 36 : 50)}px;
  line-height: ${({ condensed }) => (condensed ? 36 : 50)}px;
  text-align: center;
`

function Table<T>({
  data = [],
  columns,
  onRowClick,
  rowActionName,
  rowActions,
  icon,
  iconColor,
  headless,
  fixedLayout,
  onReorder,
  activeRowIndices,
  condensed,
  ...props
}: TableProps<T>) {
  const uid = useUniqueId()
  const standardizedColumns: Array<Column<T>> = columns.map(column => {
    if (typeof column === "string") {
      return {
        heading: column,
        cell: (dataEntry: T) => dataEntry[column],
      }
    } else {
      return column
    }
  })

  const hasIcons = Boolean(data[0] && icon && icon(data[0]))

  const handleKeyDownOnRow = React.useCallback(
    (entry, index) => (e: React.KeyboardEvent<HTMLTableRowElement>) => {
      if (!onRowClick) {
        return
      }
      switch (e.key) {
        case "Enter":
          onRowClick(entry, index)
      }
    },
    [onRowClick],
  )

  /**
   * This exists so we don't have to extract <Tr /> further
   * down, which relies on its closure.
   */
  const PseudoDraggable = React.useMemo(() => {
    if (onReorder) {
      return Draggable
    }

    return (((props: DraggableProps) => <>{props.children({} as any, {} as any)}</>) as unknown) as typeof Draggable
  }, [onReorder])

  return (
    <DragDropContext onDragEnd={onReorder || (() => {})}>
      <Container fixedLayout={fixedLayout || Boolean(onReorder)} {...props}>
        {!headless && (
          <Thead>
            <Tr active={false /* It's a heading */}>
              {hasIcons && <Th key="-1" />}
              {standardizedColumns.map((column, columnIndex) => (
                <Th
                  key={columnIndex}
                  sortable={Boolean(column.onSortClick)}
                  onClick={() => column.onSortClick && column.onSortClick(column.sortOrder === "desc" ? "asc" : "desc")}
                >
                  <ThContent sorted={Boolean(column.sortOrder)}>
                    {column.heading}
                    {column.onSortClick && !column.sortOrder && (
                      <ChevronUpDownIcon right size={10} color="color.border.disabled" />
                    )}
                    {column.sortOrder &&
                      (column.sortOrder === "desc" ? (
                        <ChevronUpIcon right size={10} color="primary" />
                      ) : (
                        <ChevronDownIcon right size={10} color="primary" />
                      ))}
                  </ThContent>
                </Th>
              ))}
              {Boolean(rowActions || (onRowClick && rowActionName)) && <Th key="infinity" />}
            </Tr>
          </Thead>
        )}
        <Droppable droppableId={uid}>
          {droppableProvided => (
            <tbody ref={droppableProvided.innerRef} {...droppableProvided.droppableProps}>
              {data.length ? (
                data.map((dataEntry, dataEntryIndex) => {
                  /*
                   Because of how border-collapse works, we need a different border color for this TD
                   and the subsequent TD if the current row is "active"
                  */
                  const shouldTdHaveColoredBorders = activeRowIndices
                    ? activeRowIndices.includes(dataEntryIndex) || activeRowIndices.includes(dataEntryIndex + 1)
                    : false

                  const rowAction = (() => {
                    if (!rowActions) {
                      return null
                    }
                    const dataEntryRowActions = rowActions(dataEntry)
                    return (
                      <Actions coloredBorders={shouldTdHaveColoredBorders}>
                        {Array.isArray(dataEntryRowActions) ? (
                          <CondensedActionMenu
                            items={dataEntryRowActions as ActionMenuProps["items"]}
                            condensed={condensed}
                          />
                        ) : (
                          dataEntryRowActions
                        )}
                      </Actions>
                    )
                  })()
                  return (
                    <PseudoDraggable key={dataEntryIndex} draggableId={String(dataEntryIndex)} index={dataEntryIndex}>
                      {(provided, snapshot) => (
                        <Tr
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          active={activeRowIndices ? activeRowIndices.includes(dataEntryIndex) : false}
                          ref={provided.innerRef}
                          isDragging={Boolean(snapshot.isDragging)}
                          onKeyDown={handleKeyDownOnRow(dataEntry, dataEntryIndex)}
                          tabIndex={onRowClick ? 0 : undefined}
                          role={onRowClick ? "button" : undefined}
                          draggable={Boolean(onReorder)}
                          clickable={Boolean(onRowClick)}
                          key={dataEntryIndex}
                          onClick={() => {
                            if (onRowClick) {
                              onRowClick(dataEntry, dataEntryIndex)
                            }
                          }}
                          condensed={condensed}
                        >
                          {hasIcons && (
                            <CellIcon coloredBorders={shouldTdHaveColoredBorders}>
                              {/** Because has `hasIcon`, it is guaranteed that the `icon` function exists */}
                              {React.createElement(icon!(dataEntry), { color: iconColor && iconColor(dataEntry) })}
                            </CellIcon>
                          )}
                          {standardizedColumns.map((column, columnIndex) => (
                            <Td coloredBorders={shouldTdHaveColoredBorders} cellWidth={column.width} key={columnIndex}>
                              {column.cell(dataEntry, dataEntryIndex)}
                            </Td>
                          ))}
                          {rowAction}
                          {onRowClick && rowActionName && (
                            <Actions coloredBorders={shouldTdHaveColoredBorders}>
                              <ActionLabel>{rowActionName}</ActionLabel>
                            </Actions>
                          )}
                        </Tr>
                      )}
                    </PseudoDraggable>
                  )
                })
              ) : (
                <Tr active={false /* It's empty */}>
                  <EmptyView coloredBorders={false} colSpan={columns.length} condensed={condensed}>
                    There are no records available
                  </EmptyView>
                </Tr>
              )}
            </tbody>
          )}
        </Droppable>
      </Container>
    </DragDropContext>
  )
}

export default Table
