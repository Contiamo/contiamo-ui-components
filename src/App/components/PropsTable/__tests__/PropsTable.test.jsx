import React from 'react';
import { shallow } from 'enzyme';

import Table from '../PropsTable';

const sampleProps = {
  name: 'Hello',
  defaultValue: 'false',
};

describe('PropsTable', () => {
  it('Should render correctly', () => {
    expect(shallow(<Table />)).toMatchSnapshot();
  });
  it('Should render correctly with props', () => {
    expect(shallow(<Table props={sampleProps} />)).toMatchSnapshot();
  });
});
