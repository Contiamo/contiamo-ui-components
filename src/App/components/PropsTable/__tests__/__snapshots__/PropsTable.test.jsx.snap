// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropsTable Should render correctly 1`] = `
<PropsTable
  className="css-16kddqz"
/>
`;

exports[`PropsTable Should render correctly with props 1`] = `
<PropsTable
  className="css-16kddqz"
  props={
    Object {
      "defaultValue": "false",
      "name": "Hello",
    }
  }
/>
`;
